library/src/main/java/org/apache/apex/malhar/lib/fs/s3output/S3BlockUpload.java

```diff
@@ -0,0 +1,490 @@
+  {
+    @Override
+    public void process(S3InitiateFileUpload.UploadFileMetadata tuple)
+    {
+      processUploadFileMetadata(tuple);
+    }
+  };
+
+  /**
+   * Convert each block of a given file into S3BlockMetaData
+   * @param tuple UploadFileMetadata
+   */
+  protected void processUploadFileMetadata(S3InitiateFileUpload.UploadFileMetadata tuple)
+  {
+    if (currentWindowId <= windowDataManager.getLargestCompletedWindow()) {
+      return;
+    }
+    long[] blocks = tuple.getFileMetadata().getBlockIds();
+    for (int i = 0; i < blocks.length; i++) {
+      blockInfo.put(blocks[i], new S3BlockMetaData(tuple.getKeyName(), tuple.getUploadId(), i + 1));
+    }
+    if (blocks.length < 1) {
+      return;
+    }
+    blockInfo.get(blocks[blocks.length - 1]).setLastBlock(true);
+  }
+
+  @Override
+  public void setup(Context.OperatorContext context)
+  {
+    processedBlocks = new HashSet<>();
+    recordData = new ArrayList<>();
+    currentWindowRecoveryState = new HashMap<>();
+    windowDataManager.setup(context);
+    s3Client = createClient();
+  }
+
+  /**
+   * Create AmazonS3 client using AWS credentials
+   * @return AmazonS3
+   */
+  protected AmazonS3 createClient()
+  {
+    AmazonS3 client = new AmazonS3Client(new BasicAWSCredentials(accessKey, secretAccessKey));
+    if (endPoint != null) {
+      client.setEndpoint(endPoint);
+    }
+    return client;
+  }
+
+  @Override
+  public void beginWindow(long windowId)
+  {
+    currentWindowId = windowId;
+    if (windowId <= windowDataManager.getLargestCompletedWindow()) {
+      replay(windowId);
+    }
+  }
+
+  /**
+   * Replay the state.
+   * @param windowId replay window Id
+   */
+  protected void replay(long windowId)
+  {
+    try {
+      @SuppressWarnings("unchecked")
+      Map<Long, UploadBlockMetadata> recoveredData = (Map<Long, UploadBlockMetadata>)windowDataManager.retrieve(windowId);
+      if (recoveredData == null) {
+        return;
+      }
+      for (Map.Entry<Long, UploadBlockMetadata> ubm: recoveredData.entrySet()) {
+        output.emit(ubm.getValue());
+        blockInfo.remove(ubm.getKey());
+      }
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public void endWindow()
+  {
+    if (recordData.size() > 0) {
+      processWaitBlocks();
+    }
+
+    for (Long blockId : processedBlocks) {
+      blockInfo.remove(blockId);
+    }
+    processedBlocks.clear();
+    if (currentWindowId > windowDataManager.getLargestCompletedWindow()) {
+      try {
+        windowDataManager.save(currentWindowRecoveryState, currentWindowId);
+      } catch (IOException e) {
+        throw new RuntimeException("Unable to save recovery", e);
+      }
+    }
+    currentWindowRecoveryState.clear();
+  }
+
+  @Override
+  public void teardown()
+  {
+    windowDataManager.teardown();
+  }
+
+  /**
+   * Process the blocks which are in wait state.
+   */
+  private void processWaitBlocks()
```

Comment:
Blocks has to wait till it receives the file meta data. I added this method call after receives the file meta data of the waited block.